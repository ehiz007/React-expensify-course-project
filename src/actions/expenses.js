import {v4 as uuidv4} from 'uuid'
import database from '../firebase/firebase'

//Expenses Actions dispatched to the reducers to work on the store

//////////////////////////////NEW EXPENSE ADDER////////////////
//ADD_EXPENSE

export const addExpense = (expense) => ({
    expense,
    type: 'ADD_EXPENSE'
})

// dispatch action and send data to fire base database

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {description = '', note = '', amount = 0, createdAt = 0} = expenseData
        const expense = {description, note, amount, createdAt}

        return database.ref('expenses').push(expense).then( (ref) => {
           
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }))
        })
    }
}

///////////////////EXPENSE REMOVER///////////////////

//Dispatches REMOVE_EXPENSE from store
export const removeExpense = ({id} = {}) =>(
    {
        id,
        type: 'REMOVE_EXPENSE'
    }
)

//Start remove Expense from store and database

export const startRemoveExpense = ({id} = {}) => {
    
    return (dispatch) => {
         return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({id}))
         })
         

    }
}

///////////////EDIT EXPENSE //////////////////////////////

//Edit expense
export const editExpense = ({id, updates} = {}) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
}) 

export const startEditExpense = ({id, updates} = {}) => {
    console.log(id)
    return (dispatch) => {
        return database.ref(`expenses/${id}`).update(updates).then((snapshot) => {
            console.log(snapshot)
            dispatch(editExpense({
                id,
                updates
            }))
        })
    }

}


///////////////SETTING STATE TO DATA READ FROM THE STORE

//dispatches set action to store 

export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSE',
    expenses
})



export const startSetExpenses = () => {
    return (dispatch) => {
        const expenses = []
        
        return database.ref('expenses').once('value').then( (snapshot) => {
            if(snapshot !== null) {
                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()})
                })
            }
            dispatch(setExpenses(expenses)) 
        })    
    }
}






//Old Add expense
// export const addExpense = (
//     {
//         description = '',
//         note = '',
//         amount = 0,
//         createdAt= 0} = {}
//     ) => ({
//             expense:{
//                 id: uuidv4(),
//                 description,
//                 note,
//                 amount,
//                 createdAt
//             },
//             type: 'ADD_EXPENSE'
//         }
// )