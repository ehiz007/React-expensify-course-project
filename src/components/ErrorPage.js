import React from 'react'
import {Link} from 'react-router-dom'

const ErrorPage = () => (
    <div>
        <p>404! Page not found</p>
        <Link to='/'>Home</Link>
    </div>
)

export default ErrorPage