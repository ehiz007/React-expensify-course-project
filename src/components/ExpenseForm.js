import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'
import 'react-dates/initialize'

export default class ExpenseForm extends React.Component { 
    constructor(props) {
        super(props)
        this.state = {
            description:  props.expense ? props.expense.description : '',
            amount:  props.expense ? props.expense.amount.toString() : '',
            note:  props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error:''
        }
    };
    
    enterDescription = (e) => {
        const description = e.target.value 
        this.setState ( () => ({
            description 
        }))
    };
    changeAmount = (e) => {
        const amount = e.target.value

        if(amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    }
    changeNote = (e) => {
        const note = e.target.value
        this.setState(() => ({note}))
    }

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState( () => ({createdAt}) )
        }
        
    }

    onFocusChange = ({focused}) => {
        this.setState (() => ({
            calenderFocused : focused
        }))
    }
    
    onSubmit = (e) => {
        e.preventDefault()

        if ( !this.state.description || !this.state.amount) {
            const error = 'Please Enter description and amount!'
            this.setState(() => ({error}))
        } else {
            this.setState(() => ({error:''}))
            const expense = {
                description: this.state.description,
                createdAt: this.state.createdAt.valueOf(),
                amount: parseFloat(this.state.amount),
                note: this.state.note
            }
            this.props.onSubmit(expense)

        }
    }
    render() {
        
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type='text' 
                        value= {this.state.description}
                        autoFocus
                        placeholder='description'
                        onChange= {this.enterDescription}
                     />
                    <input type='number' 
                        value = {this.state.amount}
                        placeholder = 'amount'
                        onChange={this.changeAmount}
                    />

                    <SingleDatePicker
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = {this.state.calenderFocused}
                        onFocusChange = {this.onFocusChange}
                        numberOfMonths = {1}
                        isOutsideRange = {()=>false}
                    />
                    <textarea
                        placeholder='add a note for your expense (optional)'
                        value={this.state.note}
                        onChange = {this.changeNote}
                    ></textarea>
                    <button onClick={this.onSubmit}>Add Expense</button>
                </form>
            </div>
        )
        
    }

}