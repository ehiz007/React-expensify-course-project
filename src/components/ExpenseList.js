import React from 'react'
import {connect} from 'react-redux'
import Expense from './ExpenseListItem'
import selectExpenses from '../selectors/getVisibility'


export const ExpenseList = (props) => (
    <div>
        <h2>Expense List</h2>
        { props.expenses.length === 0 ? (<p>There are no expenses</p>) : (
            props.expenses.map( (expense,index) => (
                <Expense key={index} {...expense}/> 
                )
            )
        )}
        
    </div>
)

const mapStateToProps = (state) => (
    {
        expenses: selectExpenses(state.expenses,state.filters)
    }
)

export default connect(mapStateToProps)(ExpenseList)