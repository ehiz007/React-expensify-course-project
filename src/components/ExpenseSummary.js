import React from 'react'
import {connect} from 'react-redux'
import selectExpenses from '../selectors/getVisibility'
import ExpenseTotal from '../selectors/ExpenseTotal'
import numeral from 'numeral'


// switch between locales
// numeral.locale('ng');

export const ExpenseSummary = (props) => {
    const expense = props.expenseCount === 1 ? 'expense' : 'expenses'
    
    return (
        <div>
            <h3>you have {props.expenseCount} {expense} , totalling {numeral(props.expenseTotal).format('$0,0.00')}</h3>
        
        </div>
    )
}

const mapStateToProps = (state) => {
       const expenses = selectExpenses(state.expenses, state.filters)
       return {
           expenseCount: expenses.length,
           expenseTotal: ExpenseTotal(expenses)
       }
}


export default connect(mapStateToProps)(ExpenseSummary)