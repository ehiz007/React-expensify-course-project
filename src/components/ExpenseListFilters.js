import React from 'react'
import {connect} from 'react-redux'
import 'react-dates/initialize'
import {DateRangePicker} from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import moment from 'moment'
import {setTextFilter, setStartDate, setEndDate, sortByDate, sortByAmount} from '../actions/filters'



export class ExpenseListFilters extends React.Component  {
    constructor (props) {
        super(props)
        this.state = {
            calenderFocuse: null
        }
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    onSelectChange = (e) => {      
        if (e.target.value === 'amount') {
            this.props.sortByAmount()
        }else if (e.target.value === 'date') {
            this.props.sortByDate()
        }            
    }
    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)  
    }
    onFocusChange = (calenderFocuse) => {
        this.setState(() => ({calenderFocuse}))
    }
   render () {
   
       return (
            <div>
                <input type='text' value={this.props.filters.text} onChange = {this.onTextChange}/>
                <select value = {this.props.filters.sortBy} onChange = {this.onSelectChange}>
                    <option value='date'>By date</option>
                    <option value='amount'>By Amount</option>
                </select>
                <DateRangePicker
                    startDate = {moment(this.props.filters.startDate)}  
                    startDateId = 'startDateId'
                    endDateId = 'endDateId'
                    endDate = {moment(this.props.filters.endDate)}
                    onDatesChange = {this.onDatesChange}
                    focusedInput = {this.state.calenderFocuse}
                    onFocusChange = {this.onFocusChange}
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}
                    showClearDates={true}

                />
        
             </div>
       )
   }
}

const mapStateToProps= (state) => (
    {
        filters: state.filters
    }
)

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (value) => dispatch(setTextFilter(value)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)