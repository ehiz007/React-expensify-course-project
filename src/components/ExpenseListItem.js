import React from 'react'
import {connect} from 'react-redux'
import {removeExpense} from '../actions/expenses'
import {Link} from 'react-router-dom'

export const Expense= ( { id,description,amount,createdAt, dispatch } ) =>(
    <div  className='expense'>  
        <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
        
        <p>{amount} - {createdAt}</p>
        <button onClick = { () => {
             dispatch(removeExpense({id})) 
        }
            }>remove</button>
    </div>
)



export default connect()(Expense)


