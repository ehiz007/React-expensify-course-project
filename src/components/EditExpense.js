import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {editExpense} from '../actions/expenses'
import {removeExpense} from '../../src/actions/expenses'

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(expense, this.props.expense.id)
        this.props.history.push('/')
    }
    removeExpense = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                This is the edit page{this.props.expense.id} 
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.removeExpense}>Remove Expense</button>
            </div>
        )
    }
}



const mapPropsToState = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id),
    
})

const mapDispatchToState = (dispatch, props) => ({
    editExpense: (expense,id) => ( dispatch( editExpense({
        id,
        updates: expense
    }))),
    removeExpense: (id) => (dispatch(removeExpense({id})))
})

export default connect(mapPropsToState, mapDispatchToState)(EditExpensePage)

