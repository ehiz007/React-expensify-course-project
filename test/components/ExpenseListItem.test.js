import React from 'react'
import {Expense} from '../../src/components/ExpenseListItem'
import {shallow} from 'enzyme'
import expenses from '../expenses/expenses'
import toJSON from 'enzyme-to-json'

test('should display an individual expense', () => {
    const wrapper = shallow(<Expense expense={expenses[1]}/>)
    expect(toJSON(wrapper)).toMatchSnapshot()
})