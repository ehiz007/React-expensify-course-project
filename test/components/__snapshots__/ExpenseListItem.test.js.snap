// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display an individual expense 1`] = `
<div
  className="expense"
>
  <Link
    to="/edit/undefined"
  >
    <h3 />
  </Link>
  <p>
    â‚¦0.00
    -
    January 1st, 1970
  </p>
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
